src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,829 @@
+      if (error instanceof Error) {
+        regExpError.value = error
+      } else {
+        throw error
+      }
+    }
+  }
+
+  numberOfMatches.value = totalMatchingLineNumbers.length
+  matchingLineNumbers.value = Array.from(new Set(totalMatchingLineNumbers))
+
+  const preElement = codeBlock.value?.querySelector('.k-highlighted-code-block')
+  const codeElement = preElement?.querySelector('code')
+
+  if (preElement instanceof HTMLElement && codeElement instanceof HTMLElement) {
+    emit('matching-lines-change', {
+      preElement,
+      codeElement,
+      code: props.code,
+      language: props.language,
+      query: query.value,
+      matchingLineNumbers: matchingLineNumbers.value,
+    })
+  }
+
+  isProcessingInternally.value = false
+}
+
+function getMatchingLineNumbers(code: string, query: string, isRegExpMode: boolean): number[] {
+  if (isRegExpMode) {
+    return getMatchingLineNumbersByRegExp(code, query)
+  } else {
+    return getMatchingLineNumbersByExactMatch(code, query)
+  }
+}
+
+function getMatchingLineNumbersByExactMatch(code: string, query: string): number[] {
+  const totalMatchingLineNumbers: number[] = []
+  let startPos = 0
+
+  while (startPos < code.length) {
+    const pos = code.indexOf(query, startPos)
+
+    if (pos === -1) {
+      break
+    }
+
+    const lineNumber = code.substring(0, pos).split('\n').length
+    totalMatchingLineNumbers.push(lineNumber)
+
+    startPos = pos + 1
+  }
+
+  return totalMatchingLineNumbers
+}
+
+function getMatchingLineNumbersByRegExp(code: string, query: string): number[] {
+  const matches = code.matchAll(new RegExp(query, 'g'))
+  const totalMatchingLineNumbers: number[] = []
+
+  for (const match of Array.from(matches)) {
+    if (match.index !== undefined) {
+      const lineNumber = code.substring(0, match.index).split('\n').length
+      totalMatchingLineNumbers.push(lineNumber)
+    }
+  }
+
+  return totalMatchingLineNumbers
+}
+
+function toggleRegExpMode(): void {
+  isRegExpMode.value = !isRegExpMode.value
+
+  // Resets regexp error when toggling off regexp mode.
+  if (!isRegExpMode.value) {
+    regExpError.value = null
+  }
+}
+
+function toggleFilterMode(): void {
+  isFilterMode.value = !isFilterMode.value
+}
+
+type Command = {
+  /**
+   * Command handler.
+   */
+  trigger: (event: Event) => (Promise<void> | void)
+
+  /**
+   * The context in which triggering the command is permitted.
+   *
+   * Logic for triggering commands via shortcuts will check whether the allowed context is in the associated `KeyboardEvent`’s [`event.composedPath()`][1].
+   *
+   * [1]: https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath
+   */
+  allowedContext?: () => EventTarget | null
+
+  /**
+   * Disables the shortcut dynamically.
+   */
+  isDisabled?: () => boolean
+
+  shouldPreventDefaultAction?: boolean
+}
+
+const commands: Record<string, Command> = {
+  // TODO: This doesn’t make sense with multiple code blocks in one document.
```

Comment:
I came here to say this. Also, we'd need to document all of the keybindings if we ship this with the component.

Due to it possibly breaking if there are multiple, unsure if this should be included.

Also, do these only fire if the element is focused?

Reply:
That’s part of the work that I simply haven’t done yet as part of the scope of this PR.

The shortcuts certainly need to be documented. I lean towards not including the `/` shortcut. The others can be included as there is logic in place for scoping them to the respective code block. All of the other commands are only triggered if they respective event is fired while focus is on an element inside (or on) the code block. As such, they’re not in conflict with other code blocks on the page.

I’ve added a shortcuts section to the docs now.

Reply:
Do you think it would be useful or necessary to allow the key map to be customizable via props?

Reply:
I’ve decided that this is useful enough to add it. I realized that you’d otherwise have no way of using the search features via shortcuts *while* avoiding conflicts with any shortcuts that you might already use in your project. Now, users can customize shortcuts via `props.keyMap` and also unbind the default ones.

Reply:
> Do you think it would be useful or necessary to allow the key map to be customizable via props?

No, I don't think we want to add in all of that complexity. Again, the primary user for this component would likely be internal and I don't think we would want our developers customizing the keybindings. This could easily lead to code blocks in multiple Kong apps having different key bindings depending on the developer that implemented them.

If we're going to have key bindings in the first place, I don't believe they should be customizable so that we have the same UX throughout all of our apps.

Reply:
I’m fine with removing this again. The argument to not promote a deviation in used shortcuts is reasonable. We do lose the ability to avoid conflicts with other tools, but that can be considered a minor issue at most.